R/toolbar.R

```diff
@@ -4,25 +4,163 @@
+#' A button designed to fit well in small places such as in a [toolbar()].
+#'
+#' @examplesIf rlang::is_interactive()
+#' toolbar(
+#'   align = "right",
+#'   toolbar_input_button(id = "see", icon = icon("eye"), label = "See"),
+#'   toolbar_input_button(id = "save", label = "Save"),
+#'   toolbar_input_button(id = "edit", icon = icon("pencil"), label = "Edit", show_label = TRUE)
+#' )
+#'
+#' @param id The input ID.
+#' @param icon An icon to display in the button. If provided without
+#'   `show_label = TRUE`, only the icon will be visible.
+#' @param label The button label. Used as button text when `show_label = TRUE`,
+#'   or as an accessibility label when hidden. Also used as the default
+#'   tooltip text when `tooltip = TRUE`.
+#' @param show_label Whether to show the label text in the button. If `FALSE`
+#'   (the default), only the icon is shown (if provided). If `TRUE`, the label
+#'   text is shown alongside the icon.
+#' @param tooltip Tooltip text to display when hovering over the button. Can be:
+#'   * `TRUE` (default when `show_label = FALSE`) - shows a tooltip with the `label` text
+#'   * `FALSE` (default when `show_label = TRUE`) - no tooltip
+#'   * A character string - shows a tooltip with custom text
+#'   Defaults to `!show_label`.
+#' @param ... Additional attributes to pass to the button.
+#' @param disabled If `TRUE`, the button will not be clickable. Use
+#'   [shiny::updateActionButton()] to dynamically enable/disable the button.
+#' @param border Whether to show a border around the button.
+#'
+#' @return Returns a button suitable for use in a toolbar.
+#'
+#' @family Toolbar components
+#' @export
+
+toolbar_input_button <- function(
+  id,
+  label,
+  icon = NULL,
+  show_label = is.null(icon),
+  tooltip = !show_label,
+  ...,
+  disabled = FALSE,
+  border = FALSE
+) {
+  btn_type <-
+    if (show_label == FALSE && is.null(icon)) {
+      rlang::abort(
+        "If `show_label` is FALSE, `icon` must be provided."
+      )
+    } else if (show_label == FALSE && !is.null(icon)) {
+      "icon"
+    } else if (show_label && is.null(icon)) {
+      "label"
+    } else {
+      # Can't both be missing (label is required)
+      "both"
+    }
+
+  # For icon-only buttons, validate that label has text for accessibility
+  if (!show_label) {
+    label_text <- paste(unlist(find_characters(label)), collapse = " ")
+    label_text <- trimws(label_text)
+    # Verifies the label contains non-empty text
+    if (!nzchar(label_text)) {
+      warning(
+        "When `show_label = FALSE`, consider providing a non-empty string label
+        for accessibility."
+      )
+    }
+  }
+
+  # Create hidden label span for icon-only buttons (for aria-labelledby)
+  if (!show_label) {
+    label_id <- paste0("btn-label-", p_randomInt(1000, 10000))
+    hidden_label <- span(
+      id = label_id,
+      hidden = NA,
+      label
+    )
+  }
+
+  button <- shiny::actionButton(
+    id,
+    label = if (btn_type != "icon") label,
```

Comment:
Currently, `label` doesn't make it into the button element. We can either overwrite `label` and always add it or only add `hidden_label` when `show_label` is `TRUE`. I like the first option a little better, I don't generally like using a variable in later code that only exists if created in a branch somewhere above.

```suggestion
  if (!show_label) {
    label_id <- paste0("btn-label-", p_randomInt(1000, 10000))
    label <- span(
      id = label_id,
      hidden = NA,
      label
    )
  }

  button <- shiny::actionButton(
    id,
    label = label,
```